projects/packages/jetpack-mu-wpcom/src/features/launchpad/class-launchpad-task-lists.php

```diff
@@ -373,11 +373,26 @@ public function add_hooks_for_active_tasks( $task_list_id = null ) {
 			return;
 		}
 
-		$active_tasks = $this->get_active_tasks( $task_list_id );
-		foreach ( $active_tasks as $task ) {
-			$task_definition = $this->get_task( $task['id'] );
+		$task_list_id = $task_list_id ? $task_list_id : get_option( 'site_intent' );
+		// Sites without a `site_intent` option will not have any tasks.
+		if ( ! $task_list_id ) {
+			return;
+		}
+
+		$task_list = $this->get_task_list( $task_list_id );
+		if ( empty( $task_list ) || ! isset( $task_list['task_ids'] ) ) {
+			return;
+		}
+
+		foreach ( $task_list['task_ids'] as $task_id ) {
+			$task_definition = $this->get_task( $task_id );
 			if ( isset( $task_definition['add_listener_callback'] ) && is_callable( $task_definition['add_listener_callback'] ) ) {
-				call_user_func_array( $task_definition['add_listener_callback'], array( $task, $task_definition ) );
+				// We only need to know the built completion status if the task has an `add_listener_callback` property.
+				// Small optimization to not run `is_complete_callback` as often.
+				$task = $this->build_task( $task_definition );
```

Comment:
Isn't "cheaper" to just register the hooks, instead of doing `build_task` and calculating completion every time??

Reply:
It's hard to say in advance. Unless there are expensive `is_complete_callback` functions, simply checking the option array for an ID key is very fast. And the `stripe_connected` task was a good example of one that benefited from being able to update it directly from the 3rd party system. Then the state is in the `launchpad_checklist_tasks_statuses` option and checking is easy. That should be the golden path here probably and this indicates that `is_complete_callback` should perhaps be removed entirely.

Reply:
Looks like the only one registered right now is this one:

[`'is_complete_callback' => 'wpcom_is_domain_upsell_completed',`](https://github.com/Automattic/jetpack/blob/860d8cd213520e363e1699b2b35ebfa4f3ee9f57/projects/packages/jetpack-mu-wpcom/src/features/launchpad/launchpad.php#LL188C31-L188C63)

It does a [`wpcom_site_has_feature( 'custom-domain' )`](https://github.com/Automattic/jetpack/blob/860d8cd213520e363e1699b2b35ebfa4f3ee9f57/projects/packages/jetpack-mu-wpcom/src/features/launchpad/launchpad.php#L405), would you consider it efficient enough? will run for every api request to a site with a "site intent" right? 😅